lib/mongodb.js

```diff
@@ -942,12 +946,12 @@ MongoDB.prototype.buildWhere = function(modelName, where, options) {
 
     const prop = self.getPropertyDefinition(modelName, propName);
 
-    const isDecimal = prop &&
-      prop.mongodb && prop.mongodb.dataType &&
-      prop.mongodb.dataType.toLowerCase() === 'decimal128';
-    if (isDecimal) {
-      cond = Decimal128.fromString(cond);
-      debug('buildWhere decimal value: %s, constructor name: %s', cond, cond.constructor.name);
+    if (prop && prop.mongodb && typeof prop.mongodb.dataType === 'string') {
+      const isDecimal = prop.mongodb.dataType.toLowerCase() === 'decimal128';
+      if (isDecimal) {
+        cond = Decimal128.fromString(cond);
+        debug('buildWhere decimal value: %s, constructor name: %s', cond, cond.constructor.name);
+      }
     }
```

Comment:
I think you need to handle `prop.mongodb.dataType.toLowerCase() === 'objectid'` too, so that string properties stored as ObjectIDs are handled in `where` queries too. Please start with a unit test that fails now.

I am ok to leave this part out of scope of this pull request if you prefer, but I think it should be implemented before we can call the user story done.

Reply:
Given:

`const ObjectIdTypeRegex = /objectid/i;` 

Shouldn't `ObjectIdTypeRegex.test()` take care of things?

Reply:
> Shouldn't `ObjectIdTypeRegex.test()` take care of things?

What will happen when the value is not an ObjectID string? IIUC, in your proposal the query will return no records, because the non-ObjectID string will not match any of the ObjectID values stored in the database.

Are we fine with that behavior?

I was thinking that we should reject such requests with `400 Bad Request` error instead.

Also, what happens when `strictObjectIDCoercion` is enabled and the property is defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`. Will we correctly coerce the string value to ObjectID, so that it can match any existing values in the database?

Reply:
I got your point, I was thinking of something else.

Reply:
> > Shouldn't `ObjectIdTypeRegex.test()` take care of things?
> 
> What will happen when the value is not an ObjectID string? IIUC, in your proposal the query will return no records, because the non-ObjectID string will not match any of the ObjectID values stored in the database.
> 
> Are we fine with that behavior?
> 
> I was thinking that we should reject such requests with `400 Bad Request` error instead.
> 
> Also, what happens when `strictObjectIDCoercion` is enabled and the property is defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`. Will we correctly coerce the string value to ObjectID, so that it can match any existing values in the database?

I misunderstood this, sorry. Rejecting non-objectID string makes sense to me (I thought this was the behaviour). Also, for the case when strict coercion flag is enabled, we should coerce string values to ObjectID, and also reject/throw an error if we can't (invalid values).

Reply:
@bajtos @hacksparrow is this something we can create a follow-up story on?

Reply:
Let's create a follow up story.

Reply:
@b-admike @hacksparrow @bajtos 
Created #545 as I believe it relates to this conversation. This commit and its related code appears to fix the issue related to string-properties-stored-as-ObjectId but does not account for a property that is an Array of ObjectID or Array of ObjectID-like-strings.